### PR TITLE
MGMT-10122: Enhance logging for package registry

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**

### CODE REVIEW

File: pkg/registry/registry.go [hunk 1]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	obj, err := r.getHeaderFromLayer(layer, "etc/driver-toolkit-release.json")
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to find file etc/driver-toolkit-release.json in image layer: %w", err)
 	}
 
 	dtk.KernelFullVersion, found, err = unstructured.NestedString(obj.Object, "KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json: found %t, err %w", found, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:126]: fixed

File: pkg/registry/registry.go [hunk 2]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.RTKernelFullVersion, found, err = unstructured.NestedString(obj.Object, "RT_KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json: found %t, err %w", found, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:131]: fixed

File: pkg/registry/registry.go [hunk 3]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.OSVersion, found, err = unstructured.NestedString(obj.Object, "RHEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RHEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RHEL_VERSION from etc/driver-toolkit-release.json: found %t, err %w", found, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:136]: fixed

File: pkg/registry/registry.go [hunk 4]
```diff
@@ -236,17 +239,17 @@ func (r *registry) GetLayerByDigest(repo string, digest string, auth []crane.Opt
 func (r *registry) getManifestStreamFromImage(image, repo string, options []crane.Option) ([]byte, error) {
 	manifest, err := crane.Manifest(image, options...)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create crane manifest from image %s: %w", image, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:242]: fixed

File: pkg/registry/registry.go [hunk 5]
```diff
@@ -191,12 +194,12 @@ func (r *registry) ReleaseImageMachineOSConfig(layer v1.Layer) (string, error) {
 func (r *registry) GetLayersDigests(ctx context.Context, image string) (string, []string, []crane.Option, error) {
 	registry, err := r.registryFromImageURL(image)
 	if err != nil {
-		return "", nil, nil, err
+		return "", nil, nil, fmt.Errorf("failure to extract registry from image %s url in GetLayersDigests: %w", image, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:197]: fixed

File: pkg/registry/registry.go [hunk 6]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
 	dtk := &DriverToolkitEntry{}
 	obj, err := r.getHeaderFromLayer(layer, "etc/driver-toolkit-release.json")
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to find file etc/driver-toolkit-release.json in image layer: %w", err)
```
Comment (ybettan) [inline pkg/registry/registry.go:121]: Let's make `etc/driver-toolkit-release.json` a `const` and use it everywhere?

File: pkg/registry/registry.go [hunk 7]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.KernelFullVersion, found, err = unstructured.NestedString(obj.Object, "KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json: found %t: %w", found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:126]:
```suggestion
		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json, found - %t: %w", found, err)
```

File: pkg/registry/registry.go [hunk 8]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.RTKernelFullVersion, found, err = unstructured.NestedString(obj.Object, "RT_KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json: found %t: %w", found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:131]:
```suggestion
		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json, found - %t: %w", found, err)
```

File: pkg/registry/registry.go [hunk 9]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.OSVersion, found, err = unstructured.NestedString(obj.Object, "RHEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RHEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RHEL_VERSION from etc/driver-toolkit-release.json: found %t: %w", found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:136]:
```suggestion
		return nil, fmt.Errorf("failed to get RHEL_VERSION from %s, found - %t: %w", dtkReleaseJSON, found, err)
```

File: pkg/registry/registry.go [hunk 10]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 	return dtk, nil
 }
 
 func (r *registry) ReleaseManifests(layer v1.Layer) (string, error) {
 	obj, err := r.getHeaderFromLayer(layer, "release-manifests/image-references")
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to find file release-manifests/image-references in image layer: %w", err)
```
Comment (ybettan) [inline pkg/registry/registry.go:144]: Let's us make `release-manifests/image-references` a const?

File: pkg/registry/registry.go [hunk 11]
```diff
@@ -115,35 +118,35 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	tags, found, err := unstructured.NestedSlice(obj.Object, "spec", "tags")
 	if !found || err != nil {
-		return "", fmt.Errorf("failed to find spec/tag in the release-manifests/image-references: err %w, found %t", err, found)
+		return "", fmt.Errorf("failed to find spec/tag in the release-manifests/image-references: found %t, err %w", found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:149]:
```suggestion
		return "", fmt.Errorf("failed to find spec.tag in the release-manifests/image-references, found - %t: %w", found, err)
```

File: pkg/registry/registry.go [hunk 12]
```diff
@@ -158,18 +161,18 @@ func (r *registry) ReleaseManifests(layer v1.Layer) (string, error) {
...
 	obj, err := r.getHeaderFromLayer(layer, "release-manifests/image-references")
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to find file release-manifests/image-references in image layer: %w", err)
 	}
 
 	tags, found, err := unstructured.NestedSlice(obj.Object, "spec", "tags")
 	if !found || err != nil {
-		return "", fmt.Errorf("failed to find spec/tags in release-manifests/image-references: error %w, found %t", err, found)
+		return "", fmt.Errorf("failed to find spec/tags in release-manifests/image-references: found %t, error %w", found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:175]:
```suggestion
		return "", fmt.Errorf("failed to find spec.tags in release-manifests/image-references, found - %t: %w", found, err)
```

File: pkg/registry/registry.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:126]: what is dtkReleaseJSON? we don't have such a variable.

File: pkg/registry/registry.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:131]: the same. we only have the output of the file as unstructured,not sure we want to dump it all into error stack

File: pkg/registry/registry.go (hunk 9 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:136]: same as above

File: pkg/registry/registry.go (hunk 11 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:149]: fixed

File: pkg/registry/registry.go (hunk 12 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:175]: fixed

File: pkg/registry/registry.go (hunk 6 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:121]: fixed

File: pkg/registry/registry.go (hunk 10 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go]: fixed

File: pkg/registry/registry.go (hunk 7 above)
Comment (ybettan) [reply pkg/registry/registry.go]: The one you just created [here](https://github.com/openshift/special-resource-operator/pull/185#discussion_r863703634)

Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go]: still don't understand what do you mean by dtkReleaseJSON. where is it created? i don't see it in the code. We only have an unstructured object

File: pkg/registry/registry.go [hunk 13]
```diff
@@ -113,37 +118,37 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
+	obj, err := r.getHeaderFromLayer(layer, driverToolkitJSONFile)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to find file %s in image layer: %w", driverToolkitJSONFile, err)
 	}
 
 	dtk.KernelFullVersion, found, err = unstructured.NestedString(obj.Object, "KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get KERNEL_VERSION from %s: found %t: %w", driverToolkitJSONFile, found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:128]:
```suggestion
		return nil, fmt.Errorf("failed to get KERNEL_VERSION from %s, found - %t: %w", driverToolkitJSONFile, found, err)
```

File: pkg/registry/registry.go [hunk 14]
```diff
@@ -113,37 +118,37 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.RTKernelFullVersion, found, err = unstructured.NestedString(obj.Object, "RT_KERNEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from %s: found %t: %w", driverToolkitJSONFile, found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:133]:
```suggestion
		return nil, fmt.Errorf("failed to get RT_KERNEL_VERSION from %s, found - %t: %w", driverToolkitJSONFile, found, err)
```

File: pkg/registry/registry.go [hunk 15]
```diff
@@ -113,37 +118,37 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
 	}
 
 	dtk.OSVersion, found, err = unstructured.NestedString(obj.Object, "RHEL_VERSION")
 	if !found || err != nil {
-		return nil, fmt.Errorf("failed to get RHEL_VERSION from etc/driver-toolkit-release.json: err %w, found %t", err, found)
+		return nil, fmt.Errorf("failed to get RHEL_VERSION from %s: found %t: %w", driverToolkitJSONFile, found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:138]:
```suggestion
		return nil, fmt.Errorf("failed to get RHEL_VERSION from %s, found - %t: %w", driverToolkitJSONFile, found, err)
```

File: pkg/registry/registry.go [hunk 16]
```diff
@@ -113,37 +118,37 @@ func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, e
...
+	obj, err := r.getHeaderFromLayer(layer, releaseManifestImagesRefFile)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to find file %s in image layer: %w", releaseManifestImagesRefFile, err)
 	}
 
 	tags, found, err := unstructured.NestedSlice(obj.Object, "spec", "tags")
 	if !found || err != nil {
-		return "", fmt.Errorf("failed to find spec/tag in the release-manifests/image-references: err %w, found %t", err, found)
+		return "", fmt.Errorf("failed to find spec/tag in the %s: found %t: %w", releaseManifestImagesRefFile, found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:151]:
```suggestion
		return "", fmt.Errorf("failed to find spec/tag in the %s, found - %t: %w", releaseManifestImagesRefFile, found, err)
```

File: pkg/registry/registry.go [hunk 17]
```diff
@@ -158,18 +163,18 @@ func (r *registry) ReleaseManifests(layer v1.Layer) (string, error) {
...
+	obj, err := r.getHeaderFromLayer(layer, releaseManifestImagesRefFile)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to find file %s in image layer: %w", releaseManifestImagesRefFile, err)
 	}
 
 	tags, found, err := unstructured.NestedSlice(obj.Object, "spec", "tags")
 	if !found || err != nil {
-		return "", fmt.Errorf("failed to find spec/tags in release-manifests/image-references: error %w, found %t", err, found)
+		return "", fmt.Errorf("failed to find spec/tags in %s: found %t: %w", releaseManifestImagesRefFile, found, err)
```
Comment (ybettan) [inline pkg/registry/registry.go:177]:
```suggestion
		return "", fmt.Errorf("failed to find spec/tags in %s, found - %t: %w", releaseManifestImagesRefFile, found, err)
```

File: pkg/registry/registry.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go]: fixed

File: pkg/registry/registry.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go]: fixed

File: pkg/registry/registry.go (hunk 9 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go]: fixed

File: pkg/registry/registry.go (hunk 16 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:151]: fixed

File: pkg/registry/registry.go (hunk 17 above)
Comment (yevgeny-shnaidman) [reply pkg/registry/registry.go:177]: fixed

**Pull request merged**